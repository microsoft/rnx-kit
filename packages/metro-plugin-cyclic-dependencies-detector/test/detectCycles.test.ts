import {
  CyclicDependencies,
  detectCycles,
  traverseDependencies,
} from "../src/detectCycles";
import {
  entryPoint,
  graphWithCycles,
  graphWithNoCycles,
  repoRoot,
} from "./testData";

function stripRepoRootFromPaths(
  cyclicDependencies: CyclicDependencies
): CyclicDependencies {
  const token = "/~";
  return Object.entries(cyclicDependencies).reduce<CyclicDependencies>(
    (cyclicDependencies, [modulePath, importPath]) => {
      cyclicDependencies[modulePath.replace(repoRoot, token)] = importPath.map(
        (p) => p.replace(repoRoot, token)
      );
      return cyclicDependencies;
    },
    {}
  );
}

describe("traverseDependencies()", () => {
  test("returns nothing if there are no cycles", () => {
    const cyclicDependencies = traverseDependencies(
      entryPoint,
      graphWithNoCycles().dependencies,
      {}
    );
    expect(cyclicDependencies).toEqual({});
  });

  test("returns import paths causing a cycle", () => {
    const cyclicDependencies = traverseDependencies(
      entryPoint,
      graphWithCycles().dependencies,
      {}
    );
    expect(stripRepoRootFromPaths(cyclicDependencies)).toMatchSnapshot();
  });

  test("returns cycles in `node_modules`", () => {
    const cyclicDependencies = traverseDependencies(
      entryPoint,
      graphWithCycles().dependencies,
      { includeNodeModules: true }
    );
    expect(stripRepoRootFromPaths(cyclicDependencies)).toMatchSnapshot();
  });
});

describe("detectCycles()", () => {
  const consoleErrorSpy = jest.spyOn(global.console, "error");
  const consoleLogSpy = jest.spyOn(global.console, "log");
  const consoleWarnSpy = jest.spyOn(global.console, "warn");

  beforeEach(() => {
    consoleErrorSpy.mockReset();
    consoleLogSpy.mockReset();
    consoleWarnSpy.mockReset();
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  test("prints nothing if all is fine", () => {
    detectCycles(entryPoint, graphWithNoCycles(), {});
    expect(consoleWarnSpy).not.toBeCalled();
    expect(consoleErrorSpy).not.toBeCalled();
  });

  test("prints if cycles are found in `node_modules`", () => {
    detectCycles(entryPoint, graphWithNoCycles(), {
      includeNodeModules: true,
      throwOnError: false,
    });
    expect(consoleWarnSpy).toBeCalledTimes(5);
    expect(consoleErrorSpy).toBeCalledTimes(1);
  });

  test("prints import paths causing a cycle", () => {
    detectCycles(entryPoint, graphWithCycles(), { throwOnError: false });
    expect(consoleWarnSpy).toBeCalledTimes(4);
    expect(consoleErrorSpy).toBeCalledTimes(1);
  });

  test("prints import paths causing a cycle, including `node_modules`", () => {
    detectCycles(entryPoint, graphWithCycles(), {
      includeNodeModules: true,
      throwOnError: false,
    });
    expect(consoleWarnSpy).toBeCalledTimes(9);
    expect(consoleErrorSpy).toBeCalledTimes(1);
  });

  test("prints import paths causing a cycle with 0 context", () => {
    detectCycles(entryPoint, graphWithCycles(), {
      linesOfContext: 0,
      throwOnError: false,
    });
    expect(consoleWarnSpy).toBeCalledTimes(3);
    expect(consoleErrorSpy).toBeCalledTimes(1);
  });

  test("prints import paths causing a cycle with 0 context, including `node_modules`", () => {
    detectCycles(entryPoint, graphWithCycles(), {
      includeNodeModules: true,
      linesOfContext: 0,
      throwOnError: false,
    });
    expect(consoleWarnSpy).toBeCalledTimes(7);
    expect(consoleErrorSpy).toBeCalledTimes(1);
  });

  test("prints import paths causing a cycle with more context", () => {
    detectCycles(entryPoint, graphWithCycles(), {
      linesOfContext: 10,
      throwOnError: false,
    });
    expect(consoleWarnSpy).toBeCalledTimes(4);
    expect(consoleErrorSpy).toBeCalledTimes(1);
  });

  test("prints import paths causing a cycle with more context, including `node_modules`", () => {
    detectCycles(entryPoint, graphWithCycles(), {
      includeNodeModules: true,
      linesOfContext: 10,
      throwOnError: false,
    });
    expect(consoleWarnSpy).toBeCalledTimes(9);
    expect(consoleErrorSpy).toBeCalledTimes(1);
  });

  test("throws on cycle detection by default", () => {
    expect(() => detectCycles(entryPoint, graphWithCycles(), {})).toThrow(
      "Import cycles detected"
    );
    expect(consoleWarnSpy).toBeCalledTimes(4);
    expect(consoleErrorSpy).not.toBeCalled();
  });
});
