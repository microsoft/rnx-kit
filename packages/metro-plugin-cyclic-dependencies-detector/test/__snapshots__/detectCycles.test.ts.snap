// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`traverseDependencies() returns cycles in \`node_modules\` 1`] = `
Object {
  "/~/node_modules/react-native/index.js": Array [
    "/~/packages/test-app/lib/src/index.js",
    "/~/node_modules/react-native/index.js",
    "/~/node_modules/react-native/Libraries/ReactNative/AppRegistry.js",
    "/~/node_modules/react-native/Libraries/LogBox/LogBoxInspectorContainer.js",
  ],
  "/~/packages/test-app/lib/src/App.js": Array [
    "/~/packages/test-app/lib/src/index.js",
    "/~/packages/test-app/lib/src/App.js",
    "/~/packages/test-app/lib/src/cyclicExample.js",
  ],
}
`;

exports[`traverseDependencies() returns import paths causing a cycle 1`] = `
Object {
  "/~/packages/test-app/lib/src/App.js": Array [
    "/~/packages/test-app/lib/src/index.js",
    "/~/packages/test-app/lib/src/App.js",
    "/~/packages/test-app/lib/src/cyclicExample.js",
  ],
}
`;
